web/api/wrapper.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+// WrapperCommand represents components necessary for OAuth2l request
+type WrapperCommand struct {
+	RequestType string
+	Args
+	Credential
+}
+
+// Args type used for unmarshalled JSON
+type Args map[string]interface{}
+
+// Credential type used for storing JSON-formatted credentials
+type Credential map[string]interface{}
+
+// Execute will capture output of OAuth2l CLI using command args
+func (wc WrapperCommand) Execute() (output string, err error) {
+	// combinedArgs used to represent command arguments in an array
+	args, err := combinedArgs(wc)
+
+	if err != nil {
+		return "", err
+	}
+
+	// Check if credential is provided to include in oauth2l command
+	if wc.Credential != nil {
+		// Symlink path to memory file
+		path, err := getCredentialPath(wc.Credential)
+
+		if err != nil {
+			return "", err
+		}
+
+		args = append(args, "--credentials", path)
+	}
+
+	// Execute command and capture output
+	command := exec.Command("oauth2l", args...)
+	byteBuffer, err := command.Output()
+
+	// Convert byteBuffer to string and remove newline character
+	output = strings.TrimSuffix(string(byteBuffer), "\n")
+	
+	return output, err
+}
+
+func allocateMemFile(credential Credential) (descriptor int, err error) {
+	// Init cred with credential body
+	cred := credential["credential"]
+
+	byteArray := []byte(cred.(string))
+
+	// Create an anonymous file on the RAM, once all references to descriptor are dropped, file is released
+	descriptor, err = unix.MemfdCreate("credential", 0)
+
+	if err != nil {
+		return 0, err
+	}
+
+	// Truncate file size to match size of our credential body
+	err = unix.Ftruncate(descriptor, int64(len(byteArray)))
+	if err != nil {
+		return 0, err
+	}
+
+	// Map file to memory and assign virtual address
+	data, err := unix.Mmap(descriptor, 0, len(byteArray), unix.PROT_READ|unix.PROT_WRITE, unix.MAP_SHARED)
```

Comment:
Does "unix.PROT_WRITE" mean this is granting write access? If yes, is it possible to just do read access only? I understand that this is a in-memory virtual file, so granting write access should not result in security risk. However, this is more about improving reliability of this code. I think this credentials file is supposed to be immutable and setting it to read only can prevent it to be altered by accident.

Reply:
This is a good point, however, the copy command below is actually writing to this mapping directly. Therefore, removing the `PROT_WRITE` permission results in a `SIGSEGV` seg fault error during the copy step.

Using `PROT_READ` only can be done, I think what I would have to do is create a mapping with write access, copy the credential data into it, delete the mapping, then create a new mapping with read permissions only over the same address space. This would protect trying to write to the credential file. Do you think this change is necessary?

Reply:
SG. Yeah, if the copy command requires needs this permission, I think leaving it here should be fine. Right now everything is happening in this file, refactor is kind of overkilled. Let us keep it simple for the momrnt.